File: docs/src/main/scala/sexamples/basics/pingpongstate/Pinger.scala
package sexamples.basics.pingpongstate

import se.sics.kompics.sl._

// #header_and_port
// #header_only
class Pinger extends ComponentDefinition {
// #header_only

  val ppp = requires(PingPongPort);
// #header_and_port

  private var counter: Long = 0L;

  ctrl uponEvent {
    case _: Start => {
      trigger(Ping -> ppp);
    }
  }

  ppp uponEvent {
    case Pong => {
      counter += 1L;
      log.info(s"Got Pong #${counter}!");
      if (counter < 100L) {
        trigger(Ping -> ppp);
      } else {
        Kompics.asyncShutdown();
      }
    }
  }
// #header_and_port
// #header_only
}
// #header_only
// #header_and_port

File: docs/src/main/java/jexamples/basics/pingpongstate/Pinger.java
package jexamples.basics.pingpongstate;

import se.sics.kompics.Kompics;
import se.sics.kompics.ComponentDefinition;
import se.sics.kompics.Positive;
import se.sics.kompics.Handler;
import se.sics.kompics.Start;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

// #header_and_port
// #header_only
public class Pinger extends ComponentDefinition {
  // #header_only

  Positive<PingPongPort> ppp = requires(PingPongPort.class);
  // #header_and_port
  private long counter = 0;

  Handler<Start> startHandler =
      new Handler<Start>() {
        public void handle(Start event) {
          trigger(new Ping(), ppp);
        }
      };
  Handler<Pong> pongHandler =
      new Handler<Pong>() {
        public void handle(Pong event) {
          counter++;
          logger.info("Got Pong #{}!", counter);
          if (counter < 100) {
            trigger(new Ping(), ppp);
          } else {
            Kompics.asyncShutdown();
          }
        }
      };

  {
    subscribe(startHandler, control);
    subscribe(pongHandler, ppp);
  }
  // #header_and_port
  // #header_only
}
// #header_only
// #header_and_port
